### PR TITLE
VY-RES-479 - Configure Mail Notification

### PR DESCRIPTION
**Pull request by lf-achyutpkl**
@lf-kailashraj please review my code

### CODE REVIEW

File: src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                }
+            }
+        }
+    }
+
+    @Override
+    public String emailBuilder(Contract contract) {
+        String projectViewPageLink = Configuration.instance().getProjectPageViewPage() + contract.getProject().getId() + "/view";
+
+        contract.getContractMembers().stream().filter(e1 -> e1.getEmployee().getMiddleName().equals("NULL"))
```
Comment (bhuwang) [inline src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java]: MAJOR Move the "NULL" string literal on the left side of this string comparison. [rule](http://sonar.java.lftechnology.com:9000/coding_rules#rule_key=squid%3AS1132)

File: src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                employeeIds.add(cm.getEmployee().getId());
+            }
+        }
+
+        if (!employeeIds.isEmpty()) {
+            List<Employee> employees = this.fetchEmployees(employeeIds);
+            for (ContractMember cm : contract.getContractMembers()) {
+                for (Employee employee : employees) {
+
+                    if (cm.getEmployee().getId().equals(employee.getId())) {
```
Comment (bhuwang) [inline src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java]: MAJOR Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](http://sonar.java.lftechnology.com:9000/coding_rules#rule_key=squid%3AS134)

File: src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    @Inject
+    private Logger log;
+
+    @Override
+    public void pushNotification() {
+        // days before contract ending to send notification
+        LocalDate date = LocalDate.now().plusDays(Configuration.instance().getEndingNotificationDays());
+
+        List<Contract> contracts = projectService.findContractsEndingBefore(date);
+        for (Contract contract : contracts) {
```
Comment (bhuwang) [inline src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java]: MAJOR Reduce the total number of break and continue statements in this loop to use at most one. [rule](http://sonar.java.lftechnology.com:9000/coding_rules#rule_key=squid%3AS135)

File: src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        try {
+            // Create a default MimeMessage object.
+            MimeMessage message = new MimeMessage(session);
+
+            message.setFrom(new InternetAddress(from));
+            message.addRecipient(Message.RecipientType.TO, new InternetAddress(mailTo));
+            message.setSubject(subject);
+            message.setContent(body, "text/html");
+            Transport.send(message);
+        } catch (MessagingException mex) {
```
Comment (bhuwang) [inline src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java]: CRITICAL Either log or rethrow this exception. [rule](http://sonar.java.lftechnology.com:9000/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    private List<Employee> fetchEmployees(List<UUID> employeeIds) {
+
+        ResponseData<Employee> data = this.getEmployeeFromCore(employeeIds, new GenericType<ResponseData<Employee>>() {
+        });
+        return data.getData();
+    }
+
+    private <T> T getEmployeeFromCore(List<UUID> employeeIds, GenericType<T> entityType) throws WebApplicationException {
```
Comment (bhuwang) [inline src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java]: MINOR Remove the declaration of thrown exception 'javax.ws.rs.WebApplicationException' which is a runtime exception. [rule](http://sonar.java.lftechnology.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

File: src/main/java/com/lftechnology/vyaguta/resource/TaskScheduler.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    public void init() {
+        if (HOUR == null || MINUTE == null) {
+            throw new ParameterFormatException("Set valid time for notification");
+        }
+        timerService.createCalendarTimer(new ScheduleExpression().month("*").dayOfWeek("*").hour(HOUR).minute(MINUTE));
+
+    }
+
+    @Timeout
+    public void runTask(Timer timer) {
```
Comment (bhuwang) [inline src/main/java/com/lftechnology/vyaguta/resource/TaskScheduler.java]: MAJOR Remove this unused method parameter "timer". [rule](http://sonar.java.lftechnology.com:9000/coding_rules#rule_key=squid%3AS1172)

**Comment by bhuwang** [comment]
SonarQube analysis reported 6 issues:
* CRITICAL 1 critical
* MAJOR 4 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+                continue;
+            }
+
+            final String body = this.emailBuilder(contract);
+            this.sendMail(mailTo, body);
+        }
+    }
+
+    private void sendMail(String mailTo, String body) {
+        String from = "noreply@lftechnology.com";
```
Comment (lf-kailashraj) [inline src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java]: @lf-achyutpkl  should be in config file

File: src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        String projectViewPageLink = Configuration.instance().getProjectPageViewPage() + contract.getProject().getId() + "/view";
+
+        ClassLoaderTemplateResolver resolver = new ClassLoaderTemplateResolver();
+        resolver.setTemplateMode("HTML5");
+        resolver.setSuffix(".html");
+        TemplateEngine templateEngine = new TemplateEngine();
+        templateEngine.setTemplateResolver(resolver);
+        final Context context = new Context(Locale.ENGLISH);
+        context.setVariable("contract", contract);
+        context.setVariable("linkToProjectPageLink", projectViewPageLink);
```
Comment (lf-kailashraj) [inline src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectNotificationServiceImpl.java]: how about "linkToProjectPage"?

**Pull request merged**